common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -973,12 +974,30 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
   }
 
   public UTF8String replace(UTF8String search, UTF8String replace) {
-    if (EMPTY_UTF8.equals(search)) {
+    // This implementation loosely based on commons-lang3's StringUtils.replace().
+    if (numBytes == 0 || search.numBytes == 0) {
       return this;
     }
-    String replaced = toString().replace(
-      search.toString(), replace.toString());
-    return fromString(replaced);
+    // Find the first occurrence of the search string.
+    int start = 0;
+    int end = this.find(search, start);
+    if (end == -1) {
+      // Search string was not found, so string is unchanged.
+      return this;
+    }
+    // At least one match was found. Estimate space needed for result.
+    int increase = replace.numBytes - search.numBytes;
+    increase = increase < 0 ? 0 : increase;
+    increase *= 16;
```

Comment:
Modeled after https://github.com/apache/commons-lang/blob/999030a23c214a1fdcfc2f1464183e0c752777f5/src/main/java/org/apache/commons/lang3/StringUtils.java#L5617, which (somewhat arbitrarily?) allocates enough buffer space to handle 16 replacements without growth in case `max` is not set.